File: ruby/lib/limelight/casting_director.rb
#- Copyright © 2008-2011 8th Light, Inc. All Rights Reserved.
#- Limelight and all included source files are distributed under terms of the MIT License.

require 'limelight/player'
require 'limelight/builtin/players'

module Limelight

  # The CastingDirector is responsible for finding Players for Props within a Production.
  # Each Producer has an instance of a CastingDirector
  #
  # Users of Limelight need not be concerned with this class.
  #
  class CastingDirector

    attr_reader :cast

    def initialize()
      @cast = Module.new
      @fs = Java::limelight.Context.fs
    end

    def has_player(player_name, players_path)
      return @cast.const_defined?(player_name.camalized) || @fs.exists("#{@fs.join(players_path, player_name)}.rb")
    end

    alias :hasPlayer :has_player

    def cast_player(prop, player_name, players_path)
      player = load_player(player_name, players_path)
      Limelight::Player.cast(player, prop)
    end

    alias :castPlayer :cast_player

    private ###############################################

    def load_player(player_name, players_path)
      module_name = player_name.camalized
      if @cast.const_defined?(module_name)
        return @cast.const_get(module_name)
      else
        player_filename = "#{@fs.join(players_path, player_name)}.rb"
        src = @fs.read_text_file(player_filename)

        player = Player.new
        player.module_eval(src, player_filename)
        @cast.const_set(module_name, player)
        return player
      end
    end

  end

end

File: src/limelight/model/CastingDirector.java
//- Copyright © 2008-2011 8th Light, Inc. All Rights Reserved.
//- Limelight and all included source files are distributed under terms of the MIT License.

package limelight.model;

import limelight.Context;
import limelight.builtin.BuiltinBeacon;
import limelight.java.JavaPlayerRecruiter;
import limelight.model.api.Player;
import limelight.model.api.PlayerRecruiter;
import limelight.ui.model.PropPanel;
import limelight.ui.model.Scene;
import limelight.util.StringUtil;
import sun.swing.StringUIClientPropertyKey;

public class CastingDirector
{
  private PlayerRecruiter builtinPlayerRecruiter;

  public static CastingDirector installed()
  {
    CastingDirector playerRecruiter = new CastingDirector();
    Context.instance().castingDirector = playerRecruiter;
    return playerRecruiter;
  }

  public CastingDirector()
  {
    builtinPlayerRecruiter = new JavaPlayerRecruiter(ClassLoader.getSystemClassLoader());
  }

  public Player castRole(PropPanel panel, String playerName, PlayerRecruiter playerRecruiter)
  {
    Player result;

    final Scene scene = panel.getRoot();
    final String scenePlayersPath = Context.fs().pathTo(scene.getPath(), "players");
    result = castFrom(panel, playerName, playerRecruiter, scenePlayersPath);

    if(result == null && scene.getProduction() != null)
    {
      final String productionPlayersPath = Context.fs().pathTo(scene.getProduction().getPath(), "players");
      result = castFrom(panel, playerName, playerRecruiter, productionPlayersPath);
    }

    if(result == null)
      result = castFrom(panel, playerName, builtinPlayerRecruiter, BuiltinBeacon.getBuiltinPlayersPath());

    return result;
  }

  private Player castFrom(PropPanel panel, String playerName, PlayerRecruiter playerRecruiter, String playersPath)
  {
    String normalizedName = StringUtil.spearCase(playerName);
    if(playerRecruiter.canRecruit(normalizedName, playersPath))
    {
      final Player player = playerRecruiter.recruitPlayer(normalizedName, playersPath);
      player.cast(panel);
      panel.addPlayer(player);
      return player;
    }
    return null;
  }

  public void setBuiltinPlayerRecruiter(limelight.model.api.PlayerRecruiter director)
  {
    this.builtinPlayerRecruiter = director;
  }

  public limelight.model.api.PlayerRecruiter getBuiltinPlayerRecruiter()
  {
    return builtinPlayerRecruiter;
  }
}
